src/Validator/Constraints/PhoneNumber.php

```diff
@@ -43,54 +43,82 @@ class PhoneNumber extends Constraint
+
+        $types = $this->getTypes();
+        if (0 === count($types)) {
+            return null;
+        }
+
+        return reset($types);
+    }
+
+    public function getTypes(): array
+    {
+        if (is_array($this->type)) {
+            return $this->type;
         }
 
-        return self::ANY;
+        return [$this->type];
     }
 
-    public function getMessage()
+    public function getMessage(): ?string
     {
         if (null !== $this->message) {
             return $this->message;
         }
 
-        switch ($this->type) {
+        $types = $this->getTypes();
+        if (1 === count($types)) {
+            $typeName = $this->getTypeName($types[0]);
+
+            return "This value is not a valid $typeName.";
+        }
+
+        return 'This value is not a valid number.';
+    }
+
+    public function getTypeNames(): array
+    {
+        $types = is_array($this->type) ? $this->type : [$this->type];
+
+        $typeNames = [];
+        foreach ($types as $type) {
+            $typeNames[] = $this->getTypeName($type);
+        }
+
+        return $typeNames;
+    }
+
+    private function getTypeName(string $type): string
+    {
+        switch ($type) {
             case self::FIXED_LINE:
-                return 'This value is not a valid fixed-line number.';
+                return 'fixed-line number';
             case self::MOBILE:
-                return 'This value is not a valid mobile number.';
+                return 'mobile number';
             case self::PAGER:
-                return 'This value is not a valid pager number.';
+                return 'pager number';
             case self::PERSONAL_NUMBER:
-                return 'This value is not a valid personal number.';
+                return 'personal number';
             case self::PREMIUM_RATE:
-                return 'This value is not a valid premium-rate number.';
+                return 'premium-rate number';
             case self::SHARED_COST:
-                return 'This value is not a valid shared-cost number.';
+                return 'shared-cost number';
             case self::TOLL_FREE:
-                return 'This value is not a valid toll-free number.';
+                return 'toll-free number';
             case self::UAN:
-                return 'This value is not a valid UAN.';
+                return 'UAN';
             case self::VOIP:
-                return 'This value is not a valid VoIP number.';
+                return 'VoIP number';
             case self::VOICEMAIL:
-                return 'This value is not a valid voicemail access number.';
+                return 'voicemail access number';
+            case self::ANY:
+                return 'phone number';
         }
 
-        return 'This value is not a valid phone number.';
+        throw new \InvalidArgumentException("Unknown phone number type \"$type\".");
```

Comment:
Is it for BC reasons or can we introduce an exception in our namespace and use a named constructor?

InvalidArgument::type($type)

Reply:
I used the existing `Misd\PhoneNumberBundle\Exception\InvalidArgumentException` instead of `\InvalidArgumentException`.
I didn't used named constructor here (we'll see in another PR I guess).